File: internal/core/message/split.go
package message

import (
	"crypto/ed25519"
	"errors"
	"fmt"
	"time"
	"unicode/utf8"

	"github.com/mertenvg/open-dmcn/internal/core/crypto"
	"github.com/mertenvg/open-dmcn/dmcnpb"
	"google.golang.org/protobuf/proto"
)

// ctxMsgHeader domain-separates the header signature from any other signature.
const ctxMsgHeader = "dmcn-msg-header-v1\x00"

// snippetMax is how many bytes of a text body are previewed in the header.
const snippetMax = 140

// ErrBodyHashMismatch means a fetched body does not match the (signed) header's
// commitment — the body was tampered with or swapped.
var ErrBodyHashMismatch = errors.New("message: body does not match header body_hash")

// MessageHeader is the small, previewable part of a message: who/what/when plus a
// commitment to the body. Signed independently so a recipient can trust an inbox
// preview without downloading the body.
type MessageHeader struct {
	Version          uint32
	MessageID        [16]byte
	ThreadID         [16]byte
	SenderAddress    string
	SenderPublicKey  ed25519.PublicKey
	RecipientAddress string
	SentAt           time.Time
	Subject          string
	AttachmentCount  uint32
	BodySize         uint64
	Snippet          string
	ReplyToID        [16]byte
	BodyHash         [32]byte
	// BodyContentAddress is the CIDv1(raw/sha2-256) of the body ciphertext blob.
	// Set by EncryptSplit and covered by the header signature, so a verified header
	// commits to the exact ciphertext. Empty for pre-feature / non-split messages.
	BodyContentAddress []byte
	// To/Cc are the full recipient lists, identical across all recipients' copies
	// and visible to everyone. Bcc is populated only on the sender's own Sent
	// self-copy; every recipient copy carries an empty Bcc so a Bcc recipient is
	// never revealed. All three are covered by the header signature.
	To  []string
	Cc  []string
	Bcc []string
}

// SignedHeader wraps a MessageHeader with the sender's signature (which covers
// BodyHash, so a verified header also authenticates the eventual body).
type SignedHeader struct {
	Header          MessageHeader
	SenderSignature [64]byte
}

// MessageContent is the large part of a message: body + attachments.
type MessageContent struct {
	Body        MessageBody
	Attachments []AttachmentRecord
}

func (h *MessageHeader) toProto() *dmcnpb.MessageHeader {
	return &dmcnpb.MessageHeader{
		Version:            h.Version,
		MessageId:          h.MessageID[:],
		ThreadId:           h.ThreadID[:],
		SenderAddress:      h.SenderAddress,
		SenderPublicKey:    h.SenderPublicKey,
		RecipientAddress:   h.RecipientAddress,
		SentAt:             h.SentAt.Unix(),
		Subject:            h.Subject,
		AttachmentCount:    h.AttachmentCount,
		BodySize:           h.BodySize,
		Snippet:            h.Snippet,
		ReplyToId:          h.ReplyToID[:],
		BodyHash:           h.BodyHash[:],
		BodyContentAddress: h.BodyContentAddress,
		To:                 h.To,
		Cc:                 h.Cc,
		Bcc:                h.Bcc,
	}
}

func messageHeaderFromProto(pb *dmcnpb.MessageHeader) MessageHeader {
	h := MessageHeader{
		Version:          pb.Version,
		SenderAddress:    pb.SenderAddress,
		SenderPublicKey:  pb.SenderPublicKey,
		RecipientAddress: pb.RecipientAddress,
		SentAt:           time.Unix(pb.SentAt, 0).UTC(),
		Subject:          pb.Subject,
		AttachmentCount:  pb.AttachmentCount,
		BodySize:         pb.BodySize,
		Snippet:          pb.Snippet,
	}
	copy(h.MessageID[:], pb.MessageId)
	copy(h.ThreadID[:], pb.ThreadId)
	copy(h.ReplyToID[:], pb.ReplyToId)
	copy(h.BodyHash[:], pb.BodyHash)
	h.BodyContentAddress = pb.BodyContentAddress
	h.To = pb.To
	h.Cc = pb.Cc
	h.Bcc = pb.Bcc
	return h
}

func (c *MessageContent) toProto() *dmcnpb.MessageContent {
	pb := &dmcnpb.MessageContent{
		Body: &dmcnpb.MessageBody{
			ContentType: c.Body.ContentType,
			Content:     c.Body.Content,
		},
	}
	for _, a := range c.Attachments {
		pb.Attachments = append(pb.Attachments, &dmcnpb.AttachmentRecord{
			AttachmentId: a.AttachmentID[:],
			Filename:     a.Filename,
			ContentType:  a.ContentType,
			SizeBytes:    a.SizeBytes,
			ContentHash:  a.ContentHash[:],
			Content:      a.Content,
		})
	}
	return pb
}

func messageContentFromProto(pb *dmcnpb.MessageContent) MessageContent {
	c := MessageContent{
		Body: MessageBody{
			ContentType: pb.GetBody().GetContentType(),
			Content:     pb.GetBody().GetContent(),
		},
	}
	for _, a := range pb.Attachments {
		att := AttachmentRecord{
			Filename:    a.Filename,
			ContentType: a.ContentType,
			SizeBytes:   a.SizeBytes,
			Content:     a.Content,
		}
		copy(att.AttachmentID[:], a.AttachmentId)
		copy(att.ContentHash[:], a.ContentHash)
		c.Attachments = append(c.Attachments, att)
	}
	return c
}

// hash returns SHA-256 over the canonical serialization of the content, used as
// the header's body commitment.
func (c *MessageContent) hash() ([32]byte, error) {
	data, err := protoMarshal(c.toProto())
	if err != nil {
		return [32]byte{}, fmt.Errorf("message: hash content: %w", err)
	}
	return crypto.SHA256Hash(data), nil
}

// headerSignableBytes is ctxMsgHeader || canonical(MessageHeader).
func (sh *SignedHeader) headerSignableBytes() ([]byte, error) {
	data, err := protoMarshal(sh.Header.toProto())
	if err != nil {
		return nil, fmt.Errorf("message: marshal header: %w", err)
	}
	return append([]byte(ctxMsgHeader), data...), nil
}

// Sign sets SenderSignature over the canonical, context-separated header.
func (sh *SignedHeader) Sign(senderPriv ed25519.PrivateKey) error {
	data, err := sh.headerSignableBytes()
	if err != nil {
		return err
	}
	sig, err := crypto.Sign(senderPriv, data)
	if err != nil {
		return fmt.Errorf("message: sign header: %w", err)
	}
	copy(sh.SenderSignature[:], sig)
	return nil
}

// Verify validates the header signature against the header's sender public key.
func (sh *SignedHeader) Verify() error {
	data, err := sh.headerSignableBytes()
	if err != nil {
		return err
	}
	if err := crypto.Verify(sh.Header.SenderPublicKey, data, sh.SenderSignature[:]); err != nil {
		return ErrInvalidSignature
	}
	return nil
}

// Split derives a signed header + content from a composed PlaintextMessage: it
// computes the body hash, fills the preview fields, and signs the header.
func Split(msg *PlaintextMessage, senderPriv ed25519.PrivateKey) (*SignedHeader, *MessageContent, error) {
	content := &MessageContent{Body: msg.Body, Attachments: msg.Attachments}
	bodyHash, err := content.hash()
	if err != nil {
		return nil, nil, err
	}

	sh := &SignedHeader{Header: MessageHeader{
		Version:          msg.Version,
		MessageID:        msg.MessageID,
		ThreadID:         msg.ThreadID,
		SenderAddress:    msg.SenderAddress,
		SenderPublicKey:  msg.SenderPublicKey,
		RecipientAddress: msg.RecipientAddress,
		SentAt:           msg.SentAt,
		Subject:          msg.Subject,
		AttachmentCount:  uint32(len(msg.Attachments)),
		BodySize:         uint64(len(msg.Body.Content)),
		Snippet:          snippetOf(msg.Body),
		ReplyToID:        msg.ReplyToID,
		BodyHash:         bodyHash,
	}}
	if err := sh.Sign(senderPriv); err != nil {
		return nil, nil, err
	}
	return sh, content, nil
}

// snippetOf returns a short preview of a text body (empty for non-text). The
// result is always the longest valid-UTF-8 prefix of the first snippetMax bytes,
// so the signed header round-trips identically across implementations (a Go
// byte-slice could otherwise split a multibyte rune and a protobufjs verifier
// would re-encode it differently, breaking the signature).
func snippetOf(body MessageBody) string {
	if body.ContentType != "text/plain" {
		return ""
	}
	s := body.Content
	if len(s) > snippetMax {
		s = s[:snippetMax]
	}
	// Drop a trailing incomplete rune.
	for len(s) > 0 && !utf8.Valid(s) {
		s = s[:len(s)-1]
	}
	return string(s)
}

// EncryptSplit seals a signed header and content into a split EncryptedEnvelope,
// using one per-message CEK wrapped for each recipient (same KEM as Encrypt).
//
// The body is encrypted first so its ciphertext can be content-addressed
// (CIDv1 of body_nonce||encrypted_body||body_tag); that address is written into
// the header and the header is (re)signed with senderPriv before being encrypted,
// so the sender signature commits to the exact ciphertext blob. The same address
// is also carried in the clear on the envelope for keyless relay verification.
func EncryptSplit(sh *SignedHeader, content *MessageContent, recipients []RecipientInfo, senderPriv ed25519.PrivateKey) (*EncryptedEnvelope, error) {
	if len(recipients) == 0 {
		return nil, errors.New("message: encrypt split: at least one recipient required")
	}

	bodyBytes, err := protoMarshal(content.toProto())
	if err != nil {
		return nil, fmt.Errorf("message: encrypt split: marshal content: %w", err)
	}

	cek, err := crypto.RandomBytes(crypto.AES256KeySize)
	if err != nil {
		return nil, fmt.Errorf("message: encrypt split: CEK: %w", err)
	}

	// Body first: it is the content-addressed unit.
	bClass := selectSizeClass(uint32(len(bodyBytes)))
	bNonce, bCT, bTag, err := crypto.AESGCMEncrypt(cek, padPayload(bodyBytes, bClass))
	if err != nil {
		return nil, fmt.Errorf("message: encrypt split: body: %w", err)
	}

	addr, err := ComputeBodyContentAddress(bNonce, bCT, bTag)
	if err != nil {
		return nil, fmt.Errorf("message: encrypt split: content address: %w", err)
	}

	// Commit to the ciphertext address in the header and (re)sign — the address did
	// not exist when Split() signed the header, so this signature is the authoritative
	// one. (The earlier Split() signature is intentionally overwritten.)
	sh.Header.BodyContentAddress = addr
	if err := sh.Sign(senderPriv); err != nil {
		return nil, fmt.Errorf("message: encrypt split: re-sign header: %w", err)
	}

	headerBytes, err := protoMarshal(&dmcnpb.SignedHeader{
		Header:          sh.Header.toProto(),
		SenderSignature: sh.SenderSignature[:],
	})
	if err != nil {
		return nil, fmt.Errorf("message: encrypt split: marshal header: %w", err)
	}
	hClass := selectSizeClass(uint32(len(headerBytes)))
	hNonce, hCT, hTag, err := crypto.AESGCMEncrypt(cek, padPayload(headerBytes, hClass))
	if err != nil {
		return nil, fmt.Errorf("message: encrypt split: header: %w", err)
	}

	recs := make([]RecipientRecord, len(recipients))
	for i, r := range recipients {
		rec, err := wrapCEK(cek, r)
		if err != nil {
			return nil, fmt.Errorf("message: encrypt split: wrap CEK %d: %w", i, err)
		}
		recs[i] = rec
	}

	env := &EncryptedEnvelope{
		Version:            2,
		MessageID:          sh.Header.MessageID,
		Recipients:         recs,
		CreatedAt:          sh.Header.SentAt.Unix(),
		EncryptedHeader:    hCT,
		HeaderSizeClass:    hClass,
		EncryptedBody:      bCT,
		BodySizeClass:      bClass,
		BodyContentAddress: addr,
	}
	copy(env.HeaderNonce[:], hNonce)
	copy(env.HeaderTag[:], hTag)
	copy(env.BodyNonce[:], bNonce)
	copy(env.BodyTag[:], bTag)
	return env, nil
}

// unwrapFor finds the recipient's wrapped CEK in the envelope and unwraps it.
func (e *EncryptedEnvelope) unwrapFor(recipientPriv, recipientPub [32]byte) ([]byte, error) {
	for i := range e.Recipients {
		if e.Recipients[i].RecipientXPub == recipientPub {
			return unwrapCEK(&e.Recipients[i], recipientPriv)
		}
	}
	return nil, ErrRecipientNotFound
}

// DecryptHeader unwraps the CEK, decrypts the header, and verifies its signature.
// The returned header is safe to render as an inbox preview (its signature commits
// to the body via BodyHash).
func DecryptHeader(env *EncryptedEnvelope, recipientPriv, recipientPub [32]byte) (*SignedHeader, error) {
	if !env.IsSplit() {
		return nil, errors.New("message: envelope is not split (no header)")
	}
	cek, err := env.unwrapFor(recipientPriv, recipientPub)
	if err != nil {
		return nil, err
	}
	padded, err := crypto.AESGCMDecrypt(cek, env.HeaderNonce[:], env.EncryptedHeader, env.HeaderTag[:])
	if err != nil {
		return nil, fmt.Errorf("%w: header: %v", ErrDecryptionFailed, err)
	}
	pb := &dmcnpb.SignedHeader{}
	if err := proto.Unmarshal(unpadPayload(padded), pb); err != nil {
		return nil, fmt.Errorf("%w: header unmarshal: %v", ErrDecryptionFailed, err)
	}
	sh := &SignedHeader{Header: messageHeaderFromProto(pb.GetHeader())}
	copy(sh.SenderSignature[:], pb.GetSenderSignature())
	if err := sh.Verify(); err != nil {
		return nil, err
	}
	return sh, nil
}

// DecryptBody decrypts the body and verifies it against the (already-verified)
// header's BodyHash. Pass the header returned by DecryptHeader.
func DecryptBody(env *EncryptedEnvelope, header *MessageHeader, recipientPriv, recipientPub [32]byte) (*MessageContent, error) {
	if !env.IsSplit() {
		return nil, errors.New("message: envelope is not split (no body)")
	}
	cek, err := env.unwrapFor(recipientPriv, recipientPub)
	if err != nil {
		return nil, err
	}
	padded, err := crypto.AESGCMDecrypt(cek, env.BodyNonce[:], env.EncryptedBody, env.BodyTag[:])
	if err != nil {
		return nil, fmt.Errorf("%w: body: %v", ErrDecryptionFailed, err)
	}
	pb := &dmcnpb.MessageContent{}
	if err := proto.Unmarshal(unpadPayload(padded), pb); err != nil {
		return nil, fmt.Errorf("%w: body unmarshal: %v", ErrDecryptionFailed, err)
	}
	content := messageContentFromProto(pb)
	got, err := content.hash()
	if err != nil {
		return nil, err
	}
	if got != header.BodyHash {
		return nil, ErrBodyHashMismatch
	}
	// Content-address binding: the (already signature-verified) header commits to
	// the exact ciphertext blob. Recompute the address from the envelope's body
	// bytes and require it to match. Skipped when the header predates the feature
	// (empty address), where body_hash above is the sole commitment.
	if len(header.BodyContentAddress) > 0 {
		ok, err := env.bodyAddressMatches(header.BodyContentAddress)
		if err != nil {
			return nil, err
		}
		if !ok {
			return nil, ErrBodyAddressMismatch
		}
	}
	return &content, nil
}

File: cmd/dmcnd/web/src/lib/crypto/split.ts
// Split header/body message crypto (Phase 2), parity with internal/core/message/split.go.
//
// A message is split into a small, independently-signed header (preview) and a
// large body, both sealed with one per-message CEK. The recipient decrypts +
// verifies the header to render an inbox row without the body, then fetches and
// verifies the body (against the signed body_hash) on open.

import { aesGcmEncrypt, wrapCEK, selectSizeClass, padPayload, type RecipientInfo } from './encrypt';
import { aesGcmDecrypt, unwrapCEK, unpadPayload } from './decrypt';
import { signWithKey, verify } from './sign';
import {
  encodeMessageHeader,
  encodeSignedHeader,
  decodeSignedHeader,
  encodeMessageContent,
  decodeMessageContent,
  type MessageHeaderFields,
} from './protobuf';

// Domain separation tag for the header signature (matches ctxMsgHeader in Go).
const CTX_MSG_HEADER = (() => {
  const tag = new TextEncoder().encode('dmcn-msg-header-v1');
  const out = new Uint8Array(tag.length + 1); // trailing 0x00 NUL
  out.set(tag, 0);
  return out;
})();
const SNIPPET_MAX = 140;

export interface SplitEnvelope {
  version: number;
  messageId: Uint8Array;
  createdAt: number;
  recipients: Array<{
    deviceId: Uint8Array;
    recipientXPub: Uint8Array;
    ephemeralXPub: Uint8Array;
    wrappedCek: Uint8Array;
    cekNonce: Uint8Array;
    cekTag: Uint8Array;
  }>;
  encryptedHeader: Uint8Array;
  headerNonce: Uint8Array;
  headerTag: Uint8Array;
  headerSizeClass: number;
  encryptedBody: Uint8Array;
  bodyNonce: Uint8Array;
  bodyTag: Uint8Array;
  bodySizeClass: number;
  bodyContentAddress: Uint8Array; // cleartext CIDv1 of the body blob
}

export interface AttachmentInput {
  attachmentId: Uint8Array; // 16
  filename: string;
  contentType: string;
  sizeBytes: number;
  contentHash: Uint8Array; // 32, SHA-256 of content
  content: Uint8Array;
}

export interface ComposeInput {
  version: number;
  messageId: Uint8Array; // 16
  threadId: Uint8Array; // 16
  senderAddress: string;
  senderPublicKey: Uint8Array; // ed25519 public (32)
  senderSignKey: CryptoKey; // non-extractable Ed25519 handle for signing the header
  recipientAddress: string;
  // Full recipient lists (signed, visible to all). to/cc are identical across every
  // copy; bcc must be [] for recipient copies and only populated on the sender's own
  // Sent self-copy — the CALLER decides per copy, encryptSplit just forwards.
  to?: string[];
  cc?: string[];
  bcc?: string[];
  sentAt: number; // Unix seconds
  subject: string;
  bodyText: string;
  attachments?: AttachmentInput[];
  recipients: RecipientInfo[];
}

async function sha256(data: Uint8Array): Promise<Uint8Array> {
  return new Uint8Array(await crypto.subtle.digest('SHA-256', data));
}

function concat(a: Uint8Array, b: Uint8Array): Uint8Array {
  const out = new Uint8Array(a.length + b.length);
  out.set(a, 0);
  out.set(b, a.length);
  return out;
}

// bodyContentAddress is the CIDv1(raw, sha2-256) of the body blob
// (body_nonce||encrypted_body||body_tag): 0x01 0x55 0x12 0x20 || SHA-256(blob).
// Byte-for-byte parity with Go message.ComputeBodyContentAddress.
async function bodyContentAddress(nonce: Uint8Array, ciphertext: Uint8Array, tag: Uint8Array): Promise<Uint8Array> {
  const digest = await sha256(concat(concat(nonce, ciphertext), tag));
  const out = new Uint8Array(36);
  out.set([0x01, 0x55, 0x12, 0x20], 0); // CIDv1 / raw codec / sha2-256 / length 32
  out.set(digest, 4);
  return out;
}

function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
  return true;
}

// snippetOf returns the longest valid-UTF-8 prefix of the first SNIPPET_MAX bytes
// of a text body (parity with Go snippetOf — never splits a multibyte rune).
function snippetOf(contentType: string, content: Uint8Array): string {
  if (contentType !== 'text/plain') return '';
  let s = content.length > SNIPPET_MAX ? content.slice(0, SNIPPET_MAX) : content;
  while (s.length > 0) {
    try {
      return new TextDecoder('utf-8', { fatal: true }).decode(s);
    } catch {
      s = s.slice(0, s.length - 1);
    }
  }
  return '';
}

// encryptSplit builds a signed header + body sealed under one CEK and returns the
// split envelope. The caller encodes it, hashes it, and signs the hash for STORE.
export async function encryptSplit(input: ComposeInput): Promise<SplitEnvelope> {
  const bodyContent = new TextEncoder().encode(input.bodyText);
  const attachments = input.attachments ?? [];

  // Body (content) → canonical bytes → body_hash commitment.
  const contentBytes = await encodeMessageContent({
    body: { contentType: 'text/plain', content: bodyContent },
    attachments,
  });
  const bodyHash = await sha256(contentBytes);

  // Seal the body first under a fresh CEK so it can be content-addressed; the
  // address is then committed in the header and covered by the header signature.
  const cek = crypto.getRandomValues(new Uint8Array(32));

  const bClass = selectSizeClass(contentBytes.length);
  const b = await aesGcmEncrypt(cek, padPayload(contentBytes, bClass));
  const addr = await bodyContentAddress(b.nonce, b.ciphertext, b.tag);

  // Header, signed independently with the domain-separation tag. It commits to
  // both the plaintext (bodyHash) and the ciphertext blob (bodyContentAddress).
  const header: MessageHeaderFields = {
    version: input.version,
    messageId: input.messageId,
    threadId: input.threadId,
    senderAddress: input.senderAddress,
    senderPublicKey: input.senderPublicKey,
    recipientAddress: input.recipientAddress,
    sentAt: input.sentAt,
    subject: input.subject,
    attachmentCount: attachments.length,
    bodySize: bodyContent.length,
    snippet: snippetOf('text/plain', bodyContent),
    bodyHash,
    bodyContentAddress: addr,
    to: input.to,
    cc: input.cc,
    bcc: input.bcc,
  };
  const headerBytes = await encodeMessageHeader(header);
  const signature = await signWithKey(input.senderSignKey, concat(CTX_MSG_HEADER, headerBytes));
  const signedHeaderBytes = await encodeSignedHeader({ header, senderSignature: signature });

  const hClass = selectSizeClass(signedHeaderBytes.length);
  const h = await aesGcmEncrypt(cek, padPayload(signedHeaderBytes, hClass));

  const recipients = await Promise.all(input.recipients.map(r => wrapCEK(cek, r)));

  return {
    version: 2,
    messageId: input.messageId,
    createdAt: input.sentAt,
    recipients,
    encryptedHeader: h.ciphertext,
    headerNonce: h.nonce,
    headerTag: h.tag,
    headerSizeClass: hClass,
    encryptedBody: b.ciphertext,
    bodyNonce: b.nonce,
    bodyTag: b.tag,
    bodySizeClass: bClass,
    bodyContentAddress: addr,
  };
}

export interface MailboxEntryLike {
  recipients: Array<{
    recipientXPub: Uint8Array;
    ephemeralXPub: Uint8Array;
    wrappedCek: Uint8Array;
    cekNonce: Uint8Array;
    cekTag: Uint8Array;
  }>;
  encryptedHeader: Uint8Array;
  headerNonce: Uint8Array;
  headerTag: Uint8Array;
}

export interface MailboxBodyLike {
  encryptedBody: Uint8Array;
  bodyNonce: Uint8Array;
  bodyTag: Uint8Array;
}

function findRecipient(
  recipients: MailboxEntryLike['recipients'],
  x25519Pub: Uint8Array
) {
  const rec = recipients.find(r => bytesEqual(new Uint8Array(r.recipientXPub), x25519Pub));
  if (!rec) throw new Error('recipient not found in envelope');
  return {
    recipientXPub: new Uint8Array(rec.recipientXPub),
    ephemeralXPub: new Uint8Array(rec.ephemeralXPub),
    wrappedCek: new Uint8Array(rec.wrappedCek),
    cekNonce: new Uint8Array(rec.cekNonce),
    cekTag: new Uint8Array(rec.cekTag),
  };
}

// decryptHeader unwraps the CEK, decrypts the header, and verifies its signature.
// The returned header is a trustworthy preview (its signature commits to bodyHash).
export async function decryptHeader(
  entry: MailboxEntryLike,
  x25519Derive: CryptoKey,
  x25519Pub: Uint8Array
): Promise<MessageHeaderFields> {
  const rec = findRecipient(entry.recipients, x25519Pub);
  const cek = await unwrapCEK(rec, x25519Derive);
  const padded = await aesGcmDecrypt(
    cek,
    new Uint8Array(entry.headerNonce),
    new Uint8Array(entry.encryptedHeader),
    new Uint8Array(entry.headerTag)
  );
  const sh = await decodeSignedHeader(unpadPayload(padded));

  const signable = concat(CTX_MSG_HEADER, await encodeMessageHeader(sh.header));
  const ok = await verify(
    new Uint8Array(sh.header.senderPublicKey),
    signable,
    new Uint8Array(sh.senderSignature)
  );
  if (!ok) throw new Error('header signature verification failed');
  return sh.header;
}

// decryptBody decrypts the body and verifies it against the (already-verified)
// header's bodyHash.
export interface DecryptedAttachment {
  filename: string;
  contentType: string;
  content: Uint8Array;
}

export async function decryptBody(
  entry: MailboxEntryLike,
  body: MailboxBodyLike,
  header: MessageHeaderFields,
  x25519Derive: CryptoKey,
  x25519Pub: Uint8Array
): Promise<{ contentType: string; content: Uint8Array; bodyText: string; attachments: DecryptedAttachment[] }> {
  const rec = findRecipient(entry.recipients, x25519Pub);
  const cek = await unwrapCEK(rec, x25519Derive);
  const padded = await aesGcmDecrypt(
    cek,
    new Uint8Array(body.bodyNonce),
    new Uint8Array(body.encryptedBody),
    new Uint8Array(body.bodyTag)
  );
  const contentBytes = unpadPayload(padded);

  const got = await sha256(contentBytes);
  if (!bytesEqual(got, new Uint8Array(header.bodyHash))) {
    throw new Error('body does not match header body_hash');
  }

  // Content-address bind: the (signature-verified) header commits to the exact
  // ciphertext blob. Skipped when the header predates the feature (empty address).
  if (header.bodyContentAddress && header.bodyContentAddress.length > 0) {
    const addr = await bodyContentAddress(
      new Uint8Array(body.bodyNonce),
      new Uint8Array(body.encryptedBody),
      new Uint8Array(body.bodyTag)
    );
    if (!bytesEqual(addr, new Uint8Array(header.bodyContentAddress))) {
      throw new Error('body does not match header body_content_address');
    }
  }

  const content = await decodeMessageContent(contentBytes);
  const raw = new Uint8Array(content.body.content);
  const attachments: DecryptedAttachment[] = (content.attachments ?? []).map(a => ({
    filename: a.filename,
    contentType: a.contentType,
    content: new Uint8Array(a.content),
  }));
  return {
    contentType: content.body.contentType,
    content: raw,
    bodyText: new TextDecoder().decode(raw),
    attachments,
  };
}
